### PR TITLE
[FLINK-3564] [table] Implement distinct() for Table API

### PR DESCRIPTION
**Pull request by twalthr**
Implements some syntactic sugar. Another step towards a SQL-like API.

### CODE REVIEW

File: flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/table/test/DistinctITCase.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+@RunWith(Parameterized.class)
+public class DistinctITCase extends MultipleProgramsTestBase {
+
+	public DistinctITCase(TestExecutionMode mode){
+		super(mode);
+	}
+
+	@Test
+	public void testAllDistinct() throws Exception {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/table/test/DistinctITCase.java]: This test does not check if `distinct` is working. The tuples of `CollectionDataSets.getSmall3TupleDataSet()` are already distinct.

File: flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/table/test/DistinctITCase.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+		String expected = "1,1,Hi\n" + "2,2,Hello\n" + "3,2,Hello world\n";
+		compareResultAsText(results, expected);
+	}
+
+	@Test
+	public void testDistinct() throws Exception {
+		ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
+		TableEnvironment tableEnv = new TableEnvironment();
+
+		DataSet<Tuple3<Integer, Long, String>> input = CollectionDataSets.getSmall3TupleDataSet(env);
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/table/test/DistinctITCase.java]: Why not use `get3TupleDataSet()` which has more duplicates?

File: flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/table/test/DistinctITCase.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	@Test
+	public void testDistinctAfterAggregate() throws Exception {
+		ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
+		TableEnvironment tableEnv = new TableEnvironment();
+
+		DataSet<Tuple5<Integer, Long, Integer, String, Long>> input = CollectionDataSets.get5TupleDataSet(env);
+
+		Table table = tableEnv.fromDataSet(input, "a, b, c, d, e");
+
+		Table distinct = table.groupBy("a").select("e").distinct();
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/table/test/DistinctITCase.java]: Isn't the result of `groupBy('a).select('e)` non-deterministic? Many database systems do not even allow this.

Btw, I looked at the implementation of `GroupedTable.select()` and it looks incorrect. It does not group but just add a selection. Either we should reject the select or select the value an arbitrary row of the group.

**Comment by twalthr** [comment]
@fhueske I corrected the tests and rebased. I will create an issue for the wrong aggregation usage.

File: docs/apis/batch/libs/table.md [hunk 4]
```diff
@@ -238,6 +238,17 @@ Table result = left.union(right);
       </td>
     </tr>
 
+    <tr>
+      <td><strong>Distinct</strong></td>
+      <td>
+        <p>Similar to a SQL DISTINCT clause. Removes duplicate values and returns only distinct (different) values.</p>
```
Comment (fhueske) [inline docs/apis/batch/libs/table.md]: We should make clear that rows with distinct value combinations are returned.
How about: "Returns rows with distinct value combinations."

File: docs/apis/batch/libs/table.md [hunk 5]
```diff
@@ -340,6 +351,18 @@ val result = left.union(right);
 {% endhighlight %}
       </td>
     </tr>
+
+    <tr>
+      <td><strong>Distinct</strong></td>
+      <td>
+        <p>Similar to a SQL DISTINCT clause. Removes duplicate values and returns only distinct (different) values.</p>
```
Comment (fhueske) [inline docs/apis/batch/libs/table.md]: same as above.

**Comment by fhueske** [comment]
I had a comment about the documentation. 
Good to merge otherwise.

**Comment by twalthr** [comment]
I changed it.

**Comment by vasia** [comment]
Merging this.

**Comment by vasia** [comment]
You can close this @twalthr!

**Pull request closed without merging**